x/evm/keeper/state_transition.go

```diff
@@ -103,48 +101,14 @@ func (k Keeper) GetHashFn(ctx sdk.Context) vm.GetHashFunc {
 			k.Logger(ctx).Error("failed to cast height to int64", "error", err)
 			return common.Hash{}
 		}
-
-		switch {
-		case ctx.BlockHeight() == h:
-			// Case 1: The requested height matches the one from the context so we can retrieve the header
-			// hash directly from the context.
-			// Note: The headerHash is only set at begin block, it will be nil in case of a query context
-			headerHash := ctx.HeaderHash()
-			if len(headerHash) != 0 {
-				return common.BytesToHash(headerHash)
-			}
-
-			// only recompute the hash if not set (eg: checkTxState)
-			contextBlockHeader := ctx.BlockHeader()
-			header, err := cmttypes.HeaderFromProto(&contextBlockHeader)
-			if err != nil {
-				k.Logger(ctx).Error("failed to cast tendermint header from proto", "error", err)
-				return common.Hash{}
-			}
-
-			headerHash = header.Hash()
-			return common.BytesToHash(headerHash)
-
-		case ctx.BlockHeight() > h:
-			// Case 2: if the chain is not the current height we need to retrieve the hash from the store for the
-			// current chain epoch. This only applies if the current height is greater than the requested height.
-			histInfo, err := k.stakingKeeper.GetHistoricalInfo(ctx, h)
-			if err != nil {
-				k.Logger(ctx).Debug("historical info not found", "height", h, "err", err.Error())
-				return common.Hash{}
-			}
-
-			header, err := cmttypes.HeaderFromProto(&histInfo.Header)
-			if err != nil {
-				k.Logger(ctx).Error("failed to cast tendermint header from proto", "error", err)
-				return common.Hash{}
-			}
-
-			return common.BytesToHash(header.Hash())
-		default:
-			// Case 3: heights greater than the current one returns an empty hash.
+		if ctx.BlockHeight() < h {
+			return common.Hash{}
+		}
+		res, err := k.signClient.Header(ctx, &h)
```

Comment:
We can't use cometbft in state machine, because it might prune historical data.

Reply:
not sure where to get since sdk doesn't get [header](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/x/staking/keeper/historical_info.go) from abci request anymore

Reply:
isn't it still stores block header in historical info?

Reply:
header is empty, we only can get sdkCtx.HeaderHash() from abci request, but no field to save in `HistoricalInfo`, but isn't this history also get prune before?

Reply:
we should store the block hashes in the evm state directly.

Reply:
do you mean save header hash to this [state](https://github.com/crypto-org-chain/ethermint/blob/develop/x/evm/types/state.pb.go#L26)?

Reply:
I checked the 0.47, the historical info is the same?

Reply:
[header](https://github.com/cosmos/cosmos-sdk/blob/release/v0.50.x/baseapp/abci.go#L709) in sdk50 doesn't contain like [Version](https://github.com/cometbft/cometbft/blob/v0.38.x/state/validation.go#L22), sdk47 can set [header](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/baseapp/baseapp.go#L422) in  [BeginBlock](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/baseapp/abci.go#L154)  and [Commit](https://github.com/cosmos/cosmos-sdk/blob/release/v0.47.x/baseapp/abci.go)

Reply:
there are a few options:
- set full header in our forked sdk
- store historical block hash in EVM state, and refactor the `GetHashFn`

Reply:
since we don't have full header info from abci request, not sure if we can add hash field in histInfo or make use of header.LastBlockID.Hash

Reply:
Store ctx.HeaderHash() in evm state?